### PR TITLE
Add support for thymeleaf context-relative urls

### PR DESCRIPTION
**Pull request by gasches**
See https://github.com/thymeleaf/thymeleaf/issues/458

### CODE REVIEW

File: views/src/main/java/io/micronaut/views/ViewsRenderer.java [hunk 1]
```diff
@@ -51,7 +52,20 @@
      * @param data     response body to render it with a view
      * @return A writable where the view will be written to.
      */
-    @Nonnull Writable render(@Nonnull String viewName, @Nullable Object data);
+    @Nonnull default Writable render(@Nonnull String viewName, @Nullable Object data) {
```
Comment (jameskleeh) [inline views/src/main/java/io/micronaut/views/ViewsRenderer.java]: Why did you change this to a default method?

File: views/src/main/java/io/micronaut/views/thymeleaf/EngineContextFactory.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Implementation of the {@link IEngineContextFactory} interface for {@link WebContext} support.
+ *
+ * @author Semyon Gashchenko
+ */
+class EngineContextFactory implements IEngineContextFactory {
```
Comment (jameskleeh) [inline views/src/main/java/io/micronaut/views/thymeleaf/EngineContextFactory.java]: This should probably be a bean

File: views/src/main/java/io/micronaut/views/thymeleaf/LinkBuilder.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.util.Map;
+
+/**
+ * Implementation of {@link org.thymeleaf.linkbuilder.ILinkBuilder} interface for {@link WebContext}
+ * support.
+ *
+ * @author Semyon Gashchenko
+ */
+class LinkBuilder extends StandardLinkBuilder {
```
Comment (jameskleeh) [inline views/src/main/java/io/micronaut/views/thymeleaf/LinkBuilder.java]: This should probably be a bean

File: views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafFactory.java [hunk 4]
```diff
@@ -68,6 +68,8 @@ public AbstractConfigurableTemplateResolver templateResolver(ViewsConfiguration
     @Singleton
     public TemplateEngine templateEngine(ITemplateResolver templateResolver) {
         TemplateEngine engine = new TemplateEngine();
+        engine.setEngineContextFactory(new EngineContextFactory());
+        engine.setLinkBuilder(new LinkBuilder());
```
Comment (jameskleeh) [inline views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafFactory.java]: These items should be injected instead of directly instantiated

File: views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafViewsRenderer.java [hunk 5]
```diff
@@ -88,10 +87,11 @@ public ThymeleafViewsRenderer(AbstractConfigurableTemplateResolver templateResol
 
     @Override
     @Nonnull
-    public Writable render(@Nonnull String viewName, @Nullable Object data) {
+    public Writable render(@Nonnull String viewName, @Nullable Object data,
```
Comment (jameskleeh) [inline views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafViewsRenderer.java]: The original render method should be left as is. You could create a common protected method where you pass in the context

File: views/src/main/java/io/micronaut/views/thymeleaf/WebEngineContext.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Implementation of the {@link org.thymeleaf.context.IEngineContext} interface for web processing.
+ *
+ * @author Semyon Gashchenko
+ */
+class WebEngineContext extends EngineContext {
```
Comment (jameskleeh) [inline views/src/main/java/io/micronaut/views/thymeleaf/WebEngineContext.java]: Any particular reason why everything is package-private?

File: views/src/main/java/io/micronaut/views/thymeleaf/WebContext.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Web-oriented implementation of the {@link org.thymeleaf.context.IContext} and {@link
+ * org.thymeleaf.context.IWebContext} interfaces.
+ *
+ * @author Semyon Gashchenko
+ */
+class WebContext extends AbstractContext {
```
Comment (jameskleeh) [inline views/src/main/java/io/micronaut/views/thymeleaf/WebContext.java]: Any particular reason why everything is package-private?

File: views/src/main/java/io/micronaut/views/thymeleaf/EngineContextFactory.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Override
+    public IEngineContext createEngineContext(
```
Comment (jameskleeh) [inline views/src/main/java/io/micronaut/views/thymeleaf/EngineContextFactory.java]: This implementation could be made more dry

File: views/src/main/java/io/micronaut/views/ViewsRenderer.java (hunk 1 above)
Comment (gasches) [reply views/src/main/java/io/micronaut/views/ViewsRenderer.java]: In order to not break API

File: views/src/main/java/io/micronaut/views/thymeleaf/EngineContextFactory.java (hunk 8 above)
Comment (gasches) [reply views/src/main/java/io/micronaut/views/thymeleaf/EngineContextFactory.java]: Yeah. Thanks for comment.

File: views/src/main/java/io/micronaut/views/thymeleaf/WebContext.java (hunk 7 above)
Comment (gasches) [reply views/src/main/java/io/micronaut/views/thymeleaf/WebContext.java]: I think that class may not be useful outside this package.

File: views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafFactory.java (hunk 4 above)
Comment (gasches) [reply views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafFactory.java]: Can't think of a good use outside TemplateEngine configuring.

File: views/src/main/java/io/micronaut/views/thymeleaf/LinkBuilder.java (hunk 3 above)
Comment (gasches) [reply views/src/main/java/io/micronaut/views/thymeleaf/LinkBuilder.java]: Can't think of a good use outside TemplateEngine configuring.

File: views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafViewsRenderer.java (hunk 5 above)
Comment (gasches) [reply views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafViewsRenderer.java]: Could you please explain this. Do I need to create an abstract implementation of ViewsRenderer interface?

File: views/src/main/java/io/micronaut/views/thymeleaf/WebEngineContext.java (hunk 6 above)
Comment (gasches) [reply views/src/main/java/io/micronaut/views/thymeleaf/WebEngineContext.java]: I think that class may not be useful outside this package.

File: views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafViewsRenderer.java (hunk 5 above)
Comment (jameskleeh) [reply views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafViewsRenderer.java]: @gasches The problem is that if someone at any point was manually using the `ThymeleafViewsRenderer` and calling the original method, they would get an UnsupportedOperationException now. Previously that would not happen, so this is a breaking change. The original method needs to work as it did before

File: views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafFactory.java (hunk 4 above)
Comment (jameskleeh) [reply views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafFactory.java]: @gasches Someone may want to provide their own engine context factory and override your implementation. With the code as it is that is not possible without overriding the entire template engine. Same goes for the link builder

File: views/src/main/java/io/micronaut/views/ViewsRenderer.java (hunk 1 above)
Comment (jameskleeh) [reply views/src/main/java/io/micronaut/views/ViewsRenderer.java]: This isn't in of itself a breaking change, however it's not a good idea because it makes it easy to mistakingly change other view renderers in a breaking way. The original method should stay as it was.

File: views/src/main/java/io/micronaut/views/thymeleaf/LinkBuilder.java (hunk 3 above)
Comment (jameskleeh) [reply views/src/main/java/io/micronaut/views/thymeleaf/LinkBuilder.java]: It doesn't have to have a good use outside the template engine configuring to be a bean

File: views/src/main/java/io/micronaut/views/ViewsRenderer.java (hunk 1 above)
Comment (gasches) [reply views/src/main/java/io/micronaut/views/ViewsRenderer.java]: Ok. Left original method as it is and create new one with request object parameter and delegating to orig as default implementation.

File: views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafViewsRenderer.java (hunk 5 above)
Comment (gasches) [reply views/src/main/java/io/micronaut/views/thymeleaf/ThymeleafViewsRenderer.java]: Ok. Left original implementation as it is and create new one with request parameter and WebContext.

File: views/src/main/java/io/micronaut/views/thymeleaf/WebContext.java (hunk 7 above)
Comment (jameskleeh) [reply views/src/main/java/io/micronaut/views/thymeleaf/WebContext.java]: If the `EngineContextFactory` needs to be overriden, users will need access to this class. Please make it public

File: views/src/main/java/io/micronaut/views/thymeleaf/WebEngineContext.java (hunk 6 above)
Comment (jameskleeh) [reply views/src/main/java/io/micronaut/views/thymeleaf/WebEngineContext.java]: If the EngineContextFactory needs to be overriden, users will need access to this class. Please make it public

File: views/src/main/java/io/micronaut/views/thymeleaf/WebContext.java (hunk 7 above)
Comment (gasches) [reply views/src/main/java/io/micronaut/views/thymeleaf/WebContext.java]: Done.

File: views/src/main/java/io/micronaut/views/thymeleaf/WebEngineContext.java (hunk 6 above)
Comment (gasches) [reply views/src/main/java/io/micronaut/views/thymeleaf/WebEngineContext.java]: Done.

**Comment by jameskleeh** [comment]
Thanks for your contribution!
